### PR TITLE
Simple passthrough Jobs implementation

### LINKED ISSUE #533: JavaScript API for using jobs

### PR DESCRIPTION
**Pull request by shayneczyzewski**
# Description

This PR contains a simple passthrough Jobs implementation (`performAsync` and `delay` only, no cron support), paving the way toward a "real" job queue like pg-boss. The template is written in such a way that future Job Factories can be created pretty easily.

Note: The e2e test for jobs is not included in this PR to make it easier to review (that is here: #557). Only making existing e2e tests pass is included here.

Fixes #533 

## Type of change

Please select the option(s) that is more relevant.

- [ ] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js
Comment (shayneczyzewski) [inline waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js:38]: I am using default here @sodic, but just read your issue about default vs named 😅 Worth changing? I can if that is the direction, and I defer to you all on JS 👍🏻

Comment (shayneczyzewski) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js:38]: I am also doing this in `_jobs.js` but mainly just so I can export an instance of the class.

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (Martinsos) [inline waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:38]: Cool that you captured JobFactory on type level!

File: waspc/data/Generator/templates/server/src/jobs/_jobs.js
Comment (Martinsos) [inline waspc/data/Generator/templates/server/src/jobs/_jobs.js:5]: So this is all logic, just creating a job via that factory with given perform function. Ok great!

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (Martinsos) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:61]: Maybe also rename Haskell identifier equvialently.

**Review by Martinsos — COMMENTED** [review]
@shayneczyzewski awesome, you did the whole feature :D!

Looks great to me, and our codebase finally has a Factory in it haha (is that good or bad hm :D)!

I like it altogether, I left small comments, nothing critical.

@sodic you know JS better than me so pls give it also a thorough look!

File: waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js
Comment (shayneczyzewski) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js:22]: Hey there! I am 100% rusty on the Promise side lol but I think you understood the motivation in a future comment below. I was hoping to return an actual object (not something we need to resolve) that we can have several other fields on it, perhaps thinks like `jobId` and maybe `jobStatus` or some other stuff from `pg-boss` in future.

The calling syntax (per the Issue) I was going for was:
```js
const runningJob = mySpecialJob.performAsync({ something: "here" })
runningJob.someOtherStuffInFuture
runningJob.result().then(...)
```

I did have `performAsync` and `delay` as class methods to start with, but it required me to use the same name as instance methods for things that can be called on class or chained on instance, so a bit more confusing. Martijn also suggested an immutable instance approach so figured I would give it a go and see.

Also, we need to actually kick off the function in `_performAsync`, so that is why I do not return `fn` itself. This way, when the caller invokes `performAsync` it is actually running **before** they even try to wait for the result (which they may never do). So in my design, `peformAsync` just kicks it off, and `result()` gets the result (possibly blocking for it), but we may have other info/functions on the handle.

As for `result()` vs `result` I don't have enough experience to know the JS convention, but to me it felt better to imply a function. But I defer to ya'll on that.

File: waspc/data/Generator/templates/server/src/utils.js
Comment (shayneczyzewski) [reply waspc/data/Generator/templates/server/src/utils.js:51]: I think that is fine, I can update 👍🏻 thnx ⛳

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:71]: Yeah, this does have a pretty strong assumption. I wasn't sure of the best way to handle it. The good news is it will fail pretty explicitly on any `compile` op like so:

```bash
🐝 --- Compiling wasp code... -----------------------------------------------------

wasp-cli: /home/shayne/dev/wasp/waspc/data/Generator/templates/server/src/jobs/AwesomeJobFactory.js: openBinaryFile: does not exist (No such file or directory)
```

So I think as a dev it would be pretty easy to spot and not slip through, but baking it into the TS somehow is certainly more ideal. What do you have in mind for the mapping function, something like this?

```haskell
jobFactoryFilePath :: JobFactory -> Path' (Rel d) File' 
jobFactoryFilePath PassthroughJobFactory = [relfile|src/jobs/PassthroughJobFactory.js|]
```

Yeah I think that may make it more clear at least 👍🏻 thnx

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:61]: Sounds good 👍🏻 thnx

File: waspc/examples/todoApp/ext/serverSetup.js
Comment (shayneczyzewski) [reply waspc/examples/todoApp/ext/serverSetup.js]: I just updated it to `result.then` in part since it simplified the return statement :) 👍🏻

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs]: Yeah, I agree. I like the hard coded/explicit so you are forced to handle any new factory cases, which should hopefully achieve the goal here. Thanks again for the suggestion!

File: waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js
Comment (Martinsos) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js:22]: Ok, so this is interesting discussion!

I get now that you want `result` to be promise, in some way. Makes sense to me, as you said, you can return stuff instantly next to it.

But, then you go a step further, and say that you want `result` to be a function that returns promise. That way, people need to call `result()` to even start the execution of the job, which is cool because it becomes "lazy" -> they can choose if they want to run it or not.

But, with the example from before, that seems confusing to me!
With `const runningJob = mySpecialJob.performAsync({ something: "here" })`, I would expect the `runningJob` to be running -> but actually it is not and I have to call `result()` on it to get it running, right? Why did I call `performAsync` if I never wanted to perform it?

What if we instead say that `performAsync(...)` actually performs the job, in async manner of course (which the name says :D!), and then `result` is of type `Promise`, not `fn() : Promise`? Do we lose anything at all, any flexibility? To me that sounds like equally powerful, but more natural way to do this. `result` is not lazy then -> but why would it be, if you didn't want to perform job you shouldn't have called `performAsync`.

Comment (shayneczyzewski) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js:22]: Well, not quite... they don't need to call `result` to start it, that happens in `performAsync` as intented. They only call `result` if they want a result. The fact that `fn` gets invoked inside `_performAsync` should ensure it actually starts running.

For example, you can see in the output:
```bash
Server: Custom server setup done!
Server: Kicking off Job...
Server: Inside Job bar's callback foo:  { something: 'here' }
Server: Waiting for Job result...
```

We are inside the callback _before_ the line that starts waiting for the result.

Comment (Martinsos) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js:22]: Aha ok great then -> but then there is no reason for `result` to be a function, right?

Comment (shayneczyzewski) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js:22]: We chatted and simplified it quite a bit:

```js
  performAsync(args) {
    return {
      result: sleep(this.delayMs).then(() => this.perform(args))
    }
  }
```

There wasn't a true need to clone and call a "private" version really, more overly cautious but not needed. Also the helper function was reduced seeing as how we really didn't need the conditional check. Woot thanks @Martinsos ⛳

Comment (Martinsos) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js]: :golfing_man: :D

Comment (shayneczyzewski) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js]: I need to start thinking more like "and `.then()`," like this lady :D https://youtu.be/CkdyU_eUm1U?t=14

Comment (Martinsos) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js]: Ahahaha ok this was great, I never watched that movie so this was knew to me :D. I still wonder if there was any way to stop the "and then"s!! We will never know I guess.

**Review by Martinsos — APPROVED** [review]

**Review by sodic — COMMENTED** [review]
Looks great! I didn't have enough time to check the user (Wasper) experience thoroughly so my comments mostly talk about other stuff.

File: waspc/e2e-test/ShellCommands.hs
Comment (shayneczyzewski) [reply waspc/e2e-test/ShellCommands.hs:64]: Yeah you are right, this is pretty hacky and fragile and won't work on spaces. I can add a `NOTE`. Separating it out like you have I think does help, will do thnx 👍🏻

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:41]: Hmm, when I remove `deriving Enum` it says I cannot use `enumFrom` due to missing instance. Same with `Enum.Bounded` when using `minBound`. Perhaps I misunderstood the note, however?

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:79]: Neat idea for a possible future helper here. Yeah, pulling `jobFactories` up to top-level does probably make more sense, especially for other functions to use if needed in future 👍🏻

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:77]: Yeah, good point! Here I am kind of cheating by using `C.mkTmplFdWithDstAndData` on a "non-template" just to really get it copied from the data dir, but naming them could aid in readability. I can try to tidy it up a bit, thanks! 👍🏻

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (sodic) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:41]: No no, it was a misunderstanding.

You don't need to import `GHC.Enum` to use the `Bounded` type class, it's available by default. Instead of writing `  deriving (Show, Eq, Ord, Enum, Enum.Bounded`, you can just write ` deriving (Show, Eq, Ord, Enum, Bounded)` and remove the import `import qualified GHC.Enum as Enum`.

File: waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js
Comment (sodic) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js]: You're going to hate me for commenting this after everything is merged, but I think that awaiting `job.result()` is much better than awaiting `job.result` :sweat_smile:

Although completely valid and functional, it's very unusual to have a regular object field be of type `Promise` (I don't remember ever seeing it). In my experience, libraries always use functions for accessing promises.

This convention most likely comes from the fact that a large part of the JS community views promises as a magic thing returned by asynchronous functions.

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (sodic) [inline waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:52]: One more minor thing I failed to notice during my original review. @Martinsos pointed me to the preferred order of functions in one of my previous PRs: https://github.com/wasp-lang/wasp/pull/500#discussion_r824602425. We should probably apply the same rule here.

Can we perhaps make the linter take care of this for us?

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (sodic) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs]: Hm, seems like we do this a lot, but don't have a function for copying files. We could wrap this pattern into something like `copyFileFromTemplatesDir`. @Martinsos another good first issue?

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (sodic) [inline waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:82]: Since I'm already commenting on a merged PR, I might as well post all that I can think of :smile: 

Following from [these conventions](https://github.com/wasp-lang/wasp/pull/553#discussion_r849560283), I'd probably extract this one to the top-level as well. That said, it's still mostly personal preference. Just putting it out there :)

File: waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js
Comment (Martinsos) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js]: Hmmm, well if everybody is doing so, it is not a bad argument, but still, it sounds completely redundant, right? To me, also confusing. And I would bet I somewhere dealt with an object that was Promise.
Is there a chance we can find a valid argument for using `result()`, besides everybody else doing it? Btw do you have any examples of smb else doing it?

Maybe you are referring to async functions returning promises?
That of course makes sense, they are async :D, but in a sense we are doing that, we are just being a bit smarter by returning object that had that promise inside it, instead of it all being a promise.

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (Martinsos) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:41]: Btw I don't think we need Enum any more? Or Bounded?

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (Martinsos) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs]: However we refactored this part, right?

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (Martinsos) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:52]: Hah I don't think linter can do that :D!
It is not a hard rule, it is mostly about telling a story. So as in a story, you start with more general picture, and then you dive into the details of the plot. The same idea is with the code -> file should start with high level stuff, and as you read it, you get to specific details. Like a newspaper article.

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (Martinsos) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:82]: Makes sense, although I have to admit I am still relatively situational about this. If I feel it in some way benefits from the context it gets with `where`, I leave it there. This one? Yeah, this one might be a good candidate for top level, true. Since Job Factory is kind of a standalone concept.

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (Martinsos) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs]: Yes I like the idea!
Btw we actually have `mkTmplFd :: Path' (Rel ServerTemplatesDir) File' -> FileDraft` in ServerGenerator.Common -> it does what `copyFileFromTemplatesDir` would do, but it doesn't have the best name hm. So maybe just renaming it to make it more clear.

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (sodic) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:41]: We just looked at it, it's still needed (Martin confirmed) :)

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (sodic) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs]: Same thing as [this comment](https://github.com/wasp-lang/wasp/pull/553#discussion_r850343910), Martin was thinking of something else :)

Comment (Martinsos) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs]: Yup I got a bit confused!

File: waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js
Comment (sodic) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js]: Unfortunately, I can't think of an example that uses either approach.

Let's not add a redundant function for now and see if someone will complain. In the end, you are right, there's no reason for it to exist.

Comment (shayneczyzewski) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js]: Don't worry @sodic I will never hate you for commenting after a merge to agree with me :D jk We can leave and see how it feels. As I integrate pg-boss, we may need to tweak this API a bit anyways, but we can keep it in mind. Thanks guys 👍🏻

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:41]: Yeah definitely required to derive those two, but we can change `Bounded` as you said and remove that Prelude import as well. thanks👍🏻

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs]: Ah, great thanks. Didn't notice that one. I can swap that out instead of what I was doing 👍🏻

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:52]: In general I totally agree with this principle. Here, I was thinking since they are kind of type level helpers for JobFactory keeping them closer to that declaration made sense, but they can go below their use, np.

File: waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs
Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:82]: Haha yeah these are tricky. On this one, since it was so locally used, I felt where made sense to stay inside the function using it. Hmmm, will think on this one...

Comment (shayneczyzewski) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:82]: To quickly further elaborate on my personal preferences here, if I have a function I only intend to use within one other function, but I extracted to make the original function more readable, I typically keep it local (even if it _could_ stand on its own). I guess the question is it preferred to have flat functions with many top-level, or fewer top-level with more local helpers. I am not sure on the Haskell convention here tbh, but using `where` to me it feels more aligned with the "read like a story" argument as well. And subjectively, if I see tons of top-level it becomes harder for me to know what are "meant" to be used (kind of weak, but I hope you get the idea), and what are just one-time use helpers (even if they are below the use) from a quick scanning perspective. But my gut may be way off here 🤷🏻

Comment (Martinsos) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:82]: I understand this and agree with it, I mostly think the same way! I guess the subjective think is judging if something is a helper function or not. But ok, at the end I don't think it matters much to be honest, so I would say free for all regarding this.

Comment (sodic) [reply waspc/src/Wasp/Generator/ServerGenerator/JobGenerator.hs:82]: > so I would say free for all regarding this.

Yes, I agree

File: waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js
Comment (sodic) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js]: > I will never hate you for commenting after a merge to agree with me

Haha, glad I could help!

BTW, Martin and I discussed the API 10 minutes ago and came up with several more things to keep in mind:
1. Perhaps it makes sense to return the result directly (without keeping it inside the object). Still, I understand the motivation behind the current approach (job metadata and all that).
I am not an expert in the cron/scheduled job area, so I can't precisely estimate the likeliness of encountering specific use cases. @shayneczyzewski, you can be the judge here (and we can tweak it as we go).
2. We were also talking about the best approaches to implement a more advanced API (e.g., periodic jobs) and concluded that we'll most likely want to copy other people :smile: 
More specifically, accessing and controlling jobs from JS is not at all unique to Wasp. There are probably more than a handful of libraries that do precisely that. So it would be wise to check out what some of those tried and tested APIs look like.

Comment (shayneczyzewski) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js]: Yeah, good points to keep in mind for sure! The API is already needing some tweaks for pg-boss integration, as it natively just gives you back a promise that resolves to a job id when you submit a job. If we just return that, we will need to provide some other helpers to look up status, etc. which may be fine. Result could also then be our own helper though to wait until job completion (via their API, hiding the details) but we may still want to give them a result and a job id, and other stuff perhaps? I may pull you both in tomorrow to brainstorm a bit after I have a bit more time to play with it today if that sounds ok.

Comment (Martinsos) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js]: And finally we concluded that we don't know yet, so whatever works we will probably change it anyway as @shayneczyzewski does more work on it :D! So on my side, let's go with whatever makes most sense right now and we will adapt to use cases / pg-boss.

Comment (Martinsos) [reply waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js]: Yup @shayneczyzewski soudns perfect!

**Pull request merged**
